File: leetcode/src/00637_average_of_levels_in_binary_tree/solution.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
	var path [][2]float64
	dfs(root, 0, &path)
	res := make([]float64, 0, len(path))
	for _, p := range path {
		res = append(res, p[0]/p[1])
	}
	return res
}

func dfs(root *TreeNode, level int, path *[][2]float64) {
	if root == nil {
		return
	}
	if level == len(*path) {
		*path = append(*path, [2]float64{})
	}
	(*path)[level][0] += float64(root.Val)
	(*path)[level][1] += 1.0
	dfs(root.Left, level+1, path)
	dfs(root.Right, level+1, path)
}

File: leetcode/src/00637_average_of_levels_in_binary_tree/solution.rs
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn average_of_levels(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<f64> {
        let mut q = VecDeque::new();
        if let Some(r) = &root {
            q.push_back(r.clone());
        }
        let mut res = Vec::new();
        while !q.is_empty() {
            let mut level_sum = 0.0;
            let level_cnt = q.len();
            for _ in 0..q.len() {
                if let Some(front) = q.pop_front() {
                    let front = front.borrow();
                    level_sum += front.val as f64;
                    if let Some(left) = &front.left {
                        q.push_back(left.clone());
                    }
                    if let Some(right) = &front.right {
                        q.push_back(right.clone());
                    }
                }
            }
            res.push(((level_sum / level_cnt as f64) * 100000.0).round() / 100000.0);
        }
        res
    }
}
